### PR TITLE
Expose subadmin to OCP

### PR DESCRIPTION
**Pull request by rullzer**
Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

File: lib/public/Group/ISubAdmin.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * @since 16.0.0
+ */
+interface ISubAdmin {
+
+	/**
+	 * add a SubAdmin
+	 * @param IUser $user user to be SubAdmin
+	 * @param IGroup $group group $user becomes subadmin of
+	 * @return bool
```
Comment (ChristophWurst) [inline lib/public/Group/ISubAdmin.php]: what does the returned value mean? Please document

File: lib/public/Group/ISubAdmin.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	 *
+	 * @since 16.0.0
+	 */
+	public function createSubAdmin(IUser $user, IGroup $group): bool;
+
+	/**
+	 * delete a SubAdmin
+	 * @param IUser $user the user that is the SubAdmin
+	 * @param IGroup $group the group
+	 * @return bool
```
Comment (ChristophWurst) [inline lib/public/Group/ISubAdmin.php]: same

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
Thanks!

File: lib/public/Group/ISubAdmin.php (hunk 1 above)
Comment (rullzer) [reply lib/public/Group/ISubAdmin.php]: Nothing it seems :P always true...

**Review by ChristophWurst — APPROVED** [review]
Awesome!

File: lib/public/Group/ISubAdmin.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	 */
+	public function getGroupsSubAdmins(IGroup $group): array;
+
+	/**
+	 * get all SubAdmins
+	 * @return array
+	 *
+	 * @since 16.0.0
+	 */
+	public function getAllSubAdmins(): array;
```
Comment (nickvergessen) [inline lib/public/Group/ISubAdmin.php]: not sure this one should be exposed, the result could be very big anyway

**Review by nickvergessen — APPROVED** [review]

File: lib/public/Group/ISubAdmin.php (hunk 3 above)
Comment (ChristophWurst) [reply lib/public/Group/ISubAdmin.php]: Nope. I just need the one where I can ask if user A is subadmin of user B.

**Review by MorrisJobke — APPROVED** [review]

File: lib/private/SubAdmin.php [hunk 4]
```diff
@@ -84,16 +84,14 @@ public function createSubAdmin(IUser $user, IGroup $group) {
 
 		$this->emit('\OC\SubAdmin', 'postCreateSubAdmin', [$user, $group]);
 		\OC_Hook::emit("OC_SubAdmin", "post_createSubAdmin", ["gid" => $group->getGID()]);
-		return true;
```
Comment (leonklingele-work) [inline lib/private/SubAdmin.php]: This breaks API compatibility :(

Comment (MorrisJobke) [reply lib/private/SubAdmin.php]: As it is private namespace that is totally fine. We state that only the public API is kept stable.

**Pull request merged**